lightningd/channel_control.c

```diff
@@ -128,6 +141,470 @@ void notify_feerate_change(struct lightningd *ld)
+		if (channel != cc->channel)
+			continue;
+
+		struct json_stream *response = json_stream_success(cc->cmd);
+		json_add_string(response, "message", "Splice intiated");
+		json_add_string(response, "psbt", psbt_to_b64(tmpctx, psbt));
+
+		was_pending(command_success(cc->cmd, response));
+
+		list_del(&cc->list);
+		tal_free(cc);
+	}
+}
+
+/* Channeld sends us this in response to a user's `splice_update` request */
+static void handle_splice_confirmed_update(struct lightningd *ld,
+					   struct channel *channel,
+					   const u8 *msg)
+{
+	struct splice_command *cc;
+	struct splice_command *n;
+	struct wally_psbt *psbt;
+	bool commitments_secured;
+
+	if (!fromwire_channeld_splice_confirmed_update(tmpctx,
+						      msg,
+						      &psbt,
+						      &commitments_secured)) {
+		channel_internal_error(channel,
+				       "bad splice_confirmed_update %s",
+				       tal_hex(channel, msg));
+		return;
+	}
+
+	list_for_each_safe(&ld->splice_commands, cc, n, list) {
+		if (channel != cc->channel)
+			continue;
+
+		struct json_stream *response = json_stream_success(cc->cmd);
+		json_add_string(response, "message", "Splice updated");
+		json_add_string(response, "psbt", psbt_to_b64(tmpctx, psbt));
+		json_add_bool(response, "commitments_secured", commitments_secured);
+
+		was_pending(command_success(cc->cmd, response));
+
+		list_del(&cc->list);
+		tal_free(cc);
+	}
+}
+
+/* Channeld uses this to request the funding transaction for help building the
+ * splice tx */
+static void handle_splice_lookup_tx(struct lightningd *ld,
+				    struct channel *channel,
+				    const u8 *msg)
+{
+	struct bitcoin_txid txid;
+	struct bitcoin_tx *tx;
+	u8 *outmsg;
+
+	if (!fromwire_channeld_splice_lookup_tx(msg, &txid)) {
+		channel_internal_error(channel,
+				       "bad splice_lookup_tx %s",
+				       tal_hex(channel, msg));
+		return;
+	}
+
+	tx = wallet_transaction_get(tmpctx, ld->wallet, &txid);
+
+	if (!tx) {
+		channel_internal_error(channel,
+				       "channel control unable to find txid %s",
+				       type_to_string(tmpctx,
+				       		      struct bitcoin_txid,
+				       		      &txid));
+		return;
+	}
+
+	outmsg = towire_channeld_splice_lookup_tx_result(NULL, tx);
+	subd_send_msg(channel->owner, take(outmsg));
+}
+
+/* Extra splice data we want to store for bitcoin send tx interface */
+struct send_splice_info
+{
+	struct splice_command *cc;
+	struct channel *channel;
+	const struct bitcoin_tx *final_tx;
+	u32 output_index;
+	const char *err_msg;
+};
+
+static void handle_tx_broadcast(struct send_splice_info *info)
+{
+	struct lightningd *ld = info->channel->peer->ld;
+	struct amount_sat unused;
+	struct json_stream *response;
+	struct bitcoin_txid txid;
+	u8 *tx_bytes;
+	int num_utxos;
+
+	tx_bytes = linearize_tx(tmpctx, info->final_tx);
+	bitcoin_txid(info->final_tx, &txid);
+
+	/* This might have spent UTXOs from our wallet */
+	num_utxos = wallet_extract_owned_outputs(ld->wallet,
```

Comment:
lisa todo: verify output ownership in case of "can't submit to local mempool but ends up in block"

Reply:
extreme edgecase: what if it gets confirmed but we werent able to broadcast it? does everything still work as expected. it should.